al2022.pkr.hcl

```diff
@@ -122,11 +122,22 @@ build {
     ]
   }
 
+  ### reboot worker instance to install kernel update. enable-ecs-agent-inferentia-support needs
+  ### new kernel (if there is) to be installed.
+  provisioner "shell" {
+    inline_shebang    = "/bin/sh -ex"
+    expect_disconnect = "true"
+    inline = [
+      "sudo reboot"
+    ]
+  }
+
   provisioner "shell" {
     environment_vars = [
       "AMI_TYPE=${source.name}"
     ]
-    script = "scripts/enable-ecs-agent-inferentia-support.sh"
+    pause_before = "60s" # pause for the reboot
```

Comment:
I was reading some packer docs on this:
https://www.packer.io/docs/provisioners/shell#handling-reboots and
https://www.packer.io/docs/provisioners/shell#start_retry_timeout

It sounds like maybe we would want to add `start_retry_timeout` to this step, in case the reboot takes a bit longer than 60s. I think in that case we could also reduce the `pause_before` duration. From what I understand `pause_before` gives the reboot time to get started, and then the timeout would be a polling function that would allow for a longer reboot to take place, but exit early if it's ready ahead of time.

Reply:
I understand `start_retry_timeout` differently, it's the time we wait before we attempt to start next step. The original reason I don't use these two together is since the restart is done pretty fast, we should be able to just use `pause_before` to achieve the same result together.

But I think about it again, I think the best solution (with least future changes required) is a short `pause_before` (like 10s), with a `start_retry_timeout` (like 40s), plus another `max_retries` (like 3).

I actually tried to play with these parameters for a bit, and I see even if with `pause_before=2s` and `start_retry_timeout=2s`, there are not retries needed (the reboot seems to be really fast)

```
==> amazon-ebs.al2022neu: Provisioning with shell script: /var/folders/dk/gtccbc1d173f7bhs41gpsx080000gs/T/packer-shell727376015
==> amazon-ebs.al2022neu: + sudo reboot
==> amazon-ebs.al2022neu: Pausing 2s before the next provisioner...
==> amazon-ebs.al2022neu: Provisioning with shell script: scripts/enable-ecs-agent-inferentia-support.sh
==> amazon-ebs.al2022neu: + [[ al2022neu != \a\l\2\i\n\f ]]
==> amazon-ebs.al2022neu: + [[ al2022neu != \a\l\2\0\2\2\n\e\u ]]
==> amazon-ebs.al2022neu: + cat
==> amazon-ebs.al2022neu: + sudo mv /tmp/neuron.repo /etc/yum.repos.d/neuron.repo
==> amazon-ebs.al2022neu: ++ uname -r
==> amazon-ebs.al2022neu: ++ uname -r
==> amazon-ebs.al2022neu: + sudo yum install kernel-devel-5.15.57-30.131.amzn2022.x86_64 kernel-headers-5.15.57-30.131.amzn2022.x86_64 -y
```

I also tried with no `pause_before` or `start_retry_timeout`, packer still able to handle it pretty fine.

```
==> amazon-ebs.al2022neu: Provisioning with shell script: /var/folders/dk/gtccbc1d173f7bhs41gpsx080000gs/T/packer-shell305583717
==> amazon-ebs.al2022neu: + sudo reboot
==> amazon-ebs.al2022neu: Provisioning with shell script: scripts/enable-ecs-agent-inferentia-support.sh
==> amazon-ebs.al2022neu: + [[ al2022neu != \a\l\2\i\n\f ]]
==> amazon-ebs.al2022neu: + [[ al2022neu != \a\l\2\0\2\2\n\e\u ]]
==> amazon-ebs.al2022neu: + cat
==> amazon-ebs.al2022neu: + sudo mv /tmp/neuron.repo /etc/yum.repos.d/neuron.repo
==> amazon-ebs.al2022neu: ++ uname -r
==> amazon-ebs.al2022neu: ++ uname -r
==> amazon-ebs.al2022neu: + sudo yum install kernel-devel-5.15.57-30.131.amzn2022.x86_64 kernel-headers-5.15.57-30.131.amzn2022.x86_64 -y
```

(we are using 0817 base AMI with kernel ver 5.15.57-29.131 and 0831 distribution release 5.15.57-30.131, so seeing 5.15.57-30.131 means it has been successfully rebooted)

I will use `10s` and `40s` now, considering this extra time cost (0s vs. 10s and 10s vs. 50s(with one retry)) doesn't really matter for the frequency we are using build script.

Reply:
That's awesome research!